microsoftgraph/msgraph-sdk-python-core
# Feat/request context

## Overview

This PR attaches a request context object to the native Http Request object, or Graph Request Object, to control the behavior of middleware in the Python SDK.  A request context contains data that is persisted throughout the request and includes a `ClientRequestId` property, `MiddlewareControl` property to control behavior of middleware   as well as a `FeatureUsage`  property to keep track of middleware used in making the request. The request context is also be accessible to middleware components to override the options provided during middleware instantiation.

### Demo

Sample usage by telemetry handler to add specific headers
```py
client_request_id = request.context.client_request_id  
feature_usage = request.context.feature_usage  

request.headers.update({'client-request-id': f'{client_request_id}'})  
request.headers.update({'feature-usage': '{feature_usage}'})  
request.headers.update({'sdkVersion':   
           f 'graph-python-core/{SDK_VERSION} (featureUsage=feature_usage)'})
```

### Notes

This design uses keyword arguments to pass request-specific options to middleware as opposed to a middleware options dictionary. Consequently, this means that per-request config for custom middleware is not supported for now.

## Testing Instructions

* Pull the `feat/request-context` branch
* Instantiate a client and make a call to the graph api e.g 
```py
resp = client.get('/me')
```
* Observe that a context object was attached to the original request, and has the required properties
```py
print(resp.request.context)
print(resp.request.context.client_request_id)
print(resp.request.context.feature_usage)
```

Fixes #79 
Fixes #59
Fixes #43
Fixes #17
